### ISSUE TITLE
Az.Storage does not import Azure.Core.dll

### ISSUE THREAD

**Issue by AndrewStrickland** [Azure/azure-powershell#21134]
### Description

When using the Az.Storage (5.4.0) module in PowerShell Core (7.3.3), the Azure.Core.dll file is not loaded. I.E. I am unable to reference the types in Azure.Core.dll.

What happens:
```
PS C:\Users\xUser> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.3.3
PSEdition                      Core
GitCommitId                    7.3.3
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

PS C:\Users\xUser> Get-Module Az.Storage

ModuleType Version    PreRelease Name                                ExportedCommands
---------- -------    ---------- ----                                ----------------
Script     5.4.0                 Az.Storage                          {Add-AzRmStorageContainerLegalHold, Add-AzStorage…

PS C:\Users\xUser> $blobUri = New-Object Uri -ArgumentList $uriString
PS C:\Users\xUser> $azureCredential = New-Object Azure.AzureSasCredential -ArgumentList $sasToken
New-Object: Cannot find type [Azure.AzureSasCredential]: verify that the assembly containing this type is loaded.
```

Expected (from Desktop):

```
PS C:\Users\xUser> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.22000.1335
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.22000.1335
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1


PS C:\Users\xUser> Get-Module Az.Storage

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     5.4.0      Az.Storage                          {Add-AzRmStorageContainerLegalHold, Add-AzStorageAc...


PS C:\Users\xUser> $blobUri = New-Object Uri -ArgumentList $uriString
PS C:\Users\xUser> $azureCredential = New-Object Azure.AzureSasCredential -ArgumentList $sasToken
PS C:\Users\xUser> $azureCredential

Signature
---------
sp=r&st=2023-03-07T03:56:05Z&se=2023-03-07T11:56:05Z&spr=https&sv=2021-06-08&sr=b&sig=WzKiAy...
```

This previously worked in Core as well... it has just been some time since I've needed to execute this. 7.1 maybe?

### Issue script & Debug output

```PowerShell
PowerShell 7.3.3
PS C:\Users\xUser> $DebugPreference = 'Continue'
PS C:\Users\xUser> Import-Module Az.Storage
DEBUG: Initializing ConditionalAssemblyContext. PSEdition is [Core]. PSVersion is [7.3.3].
DEBUG: Initializing ConditionalAssemblyProvider. AssemblyRootPath is [C:\Program Files\PowerShell\Modules\Az.Accounts\2.11.2\StartupScripts\..\lib].
DEBUG: Registering Az shared AssemblyLoadContext.
DEBUG: AssemblyLoadContext registered.
PS C:\Users\xUser> $blobUri = New-Object Uri -ArgumentList '<URL>'
PS C:\Users\xUser> $azureCredential = New-Object Azure.AzureSasCredential -ArgumentList '<SasToken>'
New-Object: Cannot find type [Azure.AzureSasCredential]: verify that the assembly containing this type is loaded.
PS C:\Users\xUser>
```
```


### Environment data

```PowerShell
PS C:\Users\xUser> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.3.3
PSEdition                      Core
GitCommitId                    7.3.3
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```
```


### Module versions

```PowerShell
PS C:\Users\xUser> Get-Module Az*

ModuleType Version    PreRelease Name                                ExportedCommands
---------- -------    ---------- ----                                ----------------
Script     2.11.2                Az.Accounts                         {Add-AzEnvironment, Clear-AzConfig, Clear-AzConte…
Script     5.4.0                 Az.Storage                          {Add-AzRmStorageContainerLegalHold, Add-AzStorage…
```
```


### Error output

```PowerShell
DEBUG: Got version 0 of Az
DEBUG: Got version 0 of Az.Accounts
DEBUG: 9:13:12 PM - ResolveError begin processing with ParameterSet 'AnyErrorParameterSet'.
DEBUG: 9:13:12 PM - using account id 'xUser@customDomain.com'...
DEBUG: 9:13:12 PM - [ConfigManager] Got nothing from [DisplayBreakingChangeWarning], Module = [], Cmdlet = []. Returning default value [True].
WARNING: Upcoming breaking changes in the cmdlet 'Resolve-AzError' :
The `Resolve-Error` alias will be removed in a future release.  Please change any scripts that use this alias to use `Resolve-AzError` instead.
Note : Go to https://aka.ms/azps-changewarnings for steps to suppress this breaking change warning, and other information on breaking changes in Azure PowerShell.

   HistoryId: 4

Message        : Cannot find type [Azure.AzureSasCredential]: verify that the assembly containing this type is loaded.
StackTrace     :    at System.Management.Automation.MshCommandRuntime.ThrowTerminatingError(ErrorRecord errorRecord)
Exception      : System.Management.Automation.PSArgumentException
InvocationInfo : {New-Object}
Line           : $azureCredential = New-Object Azure.AzureSasCredential -ArgumentList '<SasToken>'
Position       : At line:1 char:20
                 + … redential = New-Object Azure.AzureSasCredential -ArgumentList 'sp=r&s …
                 +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
HistoryId      : 4



DEBUG: AzureQoSEvent:  Module: Az.Accounts:2.11.2; CommandName: Resolve-AzError; PSVersion: 7.3.3; IsSuccess: True; Duration: 00:00:00.2852636
DEBUG: 9:13:12 PM - [ConfigManager] Got [True] from [EnableDataCollection], Module = [], Cmdlet = [].
DEBUG: 9:13:12 PM - ResolveError end processing.

PS C:\Users\xUser>
```
```

**Comment by blueww** [comment]
@isra-fel 

It looks this is not Az.Storage cmdlet related issue, but customer failed to New-Object on an Azure.Core defined type in Powershell core.
Would you please help to look?

BTW, I find a workaround, when New-Object with the AssemblyQualifiedName of the "Azure.AzureSasCredential" it works on my machine (PSVersion 7.2.9).

```
# Get the AssemblyQualifiedName of type Azure.AzureSasCredential
# sample: Azure.AzureSasCredential, Azure.Core, Version=1.25.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8
$assemblyQualifiedName =  (([appdomain]::currentdomain.getassemblies() | Where-Object {$_.location -like "*Azure.Core*"}).gettypes()  | ?{$_.Name -eq "AzureSasCredential"}).AssemblyQualifiedName

$azureCredential = New-Object $assemblyQualifiedName -ArgumentList $sasToken
```

**Comment by AndrewStrickland** [comment]
@blueww - Confirmed the work around does work. Thank you.

**Comment by AndrewStrickland** [comment]
@SaurabhSharma-MSFT - I confirmed that a work around existed. Is it expected behavior to now need to fully qualify objects?

**Comment by isra-fel** [comment]
A quick question to @daxian-dbw , is it a known issue / limitation that if an assembly is loaded from an ALC (Azure.Core.dll in this case), you can't use the brackets syntax `[Azure.AzureSasCredential]` or `New-Object Azure.AzureSasCredential` to reference the types defined in the assembly?

**Comment by daxian-dbw** [comment]
Sorry for the delay of response. When using the assembly qualified type name, PowerShell depends on [Type.GetType](https://learn.microsoft.com/en-us/dotnet/api/system.type.gettype?view=net-7.0#system-type-gettype(system-string-system-boolean-system-boolean)) to load the type from the specified assembly, which essentially does a `Assembly.Load` to load the assembly.

https://github.com/PowerShell/PowerShell/blob/8b964ded3626bf8dc965ebce81d3621ff457c8b1/src/System.Management.Automation/engine/parser/TypeResolver.cs#L284-L309

In this particular case, `Azure.AzureSasCredential` cannot be resolved because Az.Storage loads the assembly in a custom assembly load context (ALC), and PowerShell doesn't search assemblies loaded in a custom ALC for types, because they are supposed to be isolated.

However, the assembly qualified type name can be resolved because `Type.GetType` is somehow able to load that assembly to the default ALC, meaning such an assembly can be found in the probing path of the `Assembly.Load` logic. It's very likely that the type resolved from this way is from a different instance of the assembly, not the same one that is loaded by the Az.Storage module.

**Comment by isra-fel** [comment]
Thanks for explaining @daxian-dbw . Good to know how PowerShell loads assemblies. However I'm pretty sure the required assembly is unavailable in the default ALC. My guess is that the [binding cache did the job here.](https://learn.microsoft.com/en-us/dotnet/framework/deployment/how-the-runtime-locates-assemblies#:~:text=Checks%20whether%20the%20assembly%20name%20has%20been%20bound%20to%20before)

Nevertheless, it's [recommended by .NET to pass the long format of assembly name.](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assembly.load?view=net-8.0#:~:text=it%27s%20recommended%20to%20call%20the%20Load%20method%20by%20passing%20the%20long%20form%20of%20the%20assembly%20name)

**Comment by TomasMalecek** [comment]
Is there any workaround that would allow me to use the normal syntax, mainly the square brackets syntax for accessing types from Azure.Core, @daxian-dbw, @isra-fel? I am used to [specifying types of variables](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_variables#types-of-variables) in my scripts to avoid surprises and this issue prevents me from doing so in certain cases. A somewhat specific example can be seen in [my Stack Overflow question on this topic](https://stackoverflow.com/q/79455255/2157640).

**Comment by TomasMalecek** [comment]
Sorry, had to play with it for a bit longer. This seem to work, even though it depends on the obsolete Assembly.LoadFrom() method:

```
PS> Import-Module Az.Storage
PS> [Reflection.Assembly]::LoadFrom(([AppDomain]::CurrentDomain.GetAssemblies() | Where-Object { $_.Location -like "*Azure.Core*"}).Location)

GAC    Version        Location
---    -------        --------
False  v4.0.30319     C:\Users\Palec\Documents\PowerShell\Modules\Az.Accounts\4.0.2\lib\netstandard2.0\Azure.Core.dll

PS> [Azure.ETag]

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    ETag                                     System.ValueType
```

Tried to get rid of Assembly.LoadFrom() but strangely, this does not do the trick:
```
PS> Import-Module Az.Storage
PS> [Reflection.Assembly]::Load(([AppDomain]::CurrentDomain.GetAssemblies() | Where-Object { $_.Location -like "*Azure.Core*"}).GetName())

GAC    Version        Location
---    -------        --------
False  v4.0.30319     C:\Users\Palec\Documents\PowerShell\Modules\Az.Accounts\4.0.2\StartupScripts\..\lib\netstandard2.0\Azure.Core.dll

PS> [Azure.ETag]
InvalidOperation: Unable to find type [Azure.ETag].
```

**Comment by TomasMalecek** [comment]
Somehow, when I want to use Az.Storage cmdlets, the solution attempt presented above as working generates strange errors like:
> Cannot convert the "Status: 200, Value: " value of type "Azure.ValueResponse`1[[Azure.Storage.Queues.Models.PeekedMessage, Azure.Storage.Queues, Version=12.21.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8]]" to type "Azure.Response`1[Azure.Storage.Queues.Models.PeekedMessage]".

or
```
Line |
  28 |  [AzureStorageQueue[]]$queues = Get-AzStorageQueue -Context $Context
     |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Method not found: 'Azure.Pageable`1<Azure.Storage.Queues.Models.QueueItem> Azure.Storage.Queues.QueueServiceClient.GetQueues(Azure.Storage.Queues.Models.QueueTraits, System.String, System.Threading.CancellationToken)'.
```

Seems it does not work as intended after all. :-(

**Comment by daxian-dbw** [comment]
@TomasMalecek I reviewed and revised my answer above.
Tl;Dr -- you can use the assembly-qualified type name in this case, such as `[Azure.ETag, Azure.Core]`.

The reason you got that weird error is because you loaded another assembly into the default ALC, and after that, you have 2 assembly instances of `Azure.Core` loaded into your process. The cmdlet `Get-AzStorageQueue` is using the instance that was loaded into the `AzSharedAssemblyLoadContext` ALC, but the `$Context` was created from the other instance you loaded into the default ALC. Therefore, a type identity confliction happens which results in the error you saw.

**Comment by TomasMalecek** [comment]
Thank you @daxian-dbw for the explanation and for the workaround. It is worth mentioning that `using namespace Azure` does not help in this case and that the number of type parameters has to be specified even for closed generics (I did not realize that and I struggled to find a solution for a bit). However, the type parameters do not need assembly qualification and may rely on `using namespace` unless they suffer from the ALC issue themselves.

These work:
```
using namespace Azure
using namespace Azure.Storage.Queues.Models

[Azure.ETag, Azure.Core]
[Azure.Response`1, Azure.Core]
[Azure.Response`1[bool], Azure.Core]
[Azure.Response`1[PeekedMessage], Azure.Core]
[Azure.Response`1[[Azure.ETag, Azure.Core]], Azure.Core]
```

These do not work:
```
using namespace Azure
using namespace Azure.Storage.Queues.Models

[ETag, Azure.Core]
[Response`1, Azure.Core]
[Azure.Response[bool], Azure.Core]
[Azure.Response[[System.Boolean, System.Private.CoreLib]], Azure.Core]
```
